### ISSUE TITLE
Design errors hierarchy

### ISSUE THREAD

**Issue by mijicd** [zio/zio-keeper#5]
API design (see #2) introduced sealed errors hierarchy. Expand the draft idea by providing additional members of error hierarchy.

**Comment by pshemass** [comment]
At this point I believe we know that core functionality will be divided into three category

1. Transport
2. Serialization
3. Membership
4. Service Discovery

```
sealed abstract class Error(msg: String = "", cause: Throwable = null) extends Exception(msg, cause)

sealed abstract class TransportError(msg: String = "", cause: Throwable = null) extends Error(msg, cause)
object TransportError {
  final case class UnxpectedError(exc: Throwable)   extends TransportError(cause = exc)
  final case class RequestTimeout(timeout: Duration) extends TransportError(msg = s"Request timeout $timeout.")
  final case class BindFailed(addr: SocketAddress, exc: Throwable)
    extends TransportError(msg = s"Failed binding to address $addr.", cause = exc)
}

sealed abstract class SerializationError(msg: String) extends Error(msg = msg)
object SerializationError {
 final case class CannotSerializeTypeError(obj: AnyRef) extends SerializationError(msg = s"Cannot serialize type")
final case class CannotDeserialize(chunk: Chunk) extends SerializationError(msg = s"Cannot deserialize chunk")
}

sealed abstract class MembershipError(msg: String) extends Error(msg = msg)
object MembershipError {
  final case class HandshakeError(msg: String) extends MembershipError(msg = msg)
  final case class UnknownNodeError(nodeId: NodeId) extends MembershipError(msg = s"Node $nodeId is not part of cluster")
}

sealed abstract class DiscoveryError(msg: String) extends Error(msg = msg)
object DiscoveryError {
  //whatever service discovery 
}
````

@mijicd @mschuwalow let me know if that makes sense.

**Comment by mschuwalow** [comment]
closed by #123
